source/aws_http_credentials_provider.c

```diff
@@ -0,0 +1,612 @@
+    if (aws_retry_strategy_acquire_retry_token(
+            impl->retry_strategy, NULL, s_on_retry_token_acquired, http_query_context, HTTP_RETRY_TIMEOUT_MS)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "(id=%p): failed to acquire retry token: %s",
+            (void *)provider,
+            aws_error_debug_str(aws_last_error()));
+        goto on_error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+    s_http_query_context_destroy(http_query_context);
+    return AWS_OP_ERR;
+}
+
+static void s_on_connection_manager_shutdown(void *user_data) {
+    struct aws_credentials_provider *provider = user_data;
+
+    aws_credentials_provider_invoke_shutdown_callback(provider);
+    struct aws_credentials_provider_http_impl *impl = provider->impl;
+    impl->request_vtable->clean_up_parameters_fn(impl->request_vtable->parameters);
+    aws_mem_release(provider->allocator, provider->impl);
+    aws_mem_release(provider->allocator, provider);
+}
+
+static void s_credentials_provider_http_destroy(struct aws_credentials_provider *provider) {
+
+    struct aws_credentials_provider_http_impl *impl = provider->impl;
+    if (impl == NULL) {
+        return;
+    }
+    aws_string_destroy(impl->endpoint);
+    aws_string_destroy(impl->account_id);
+    aws_retry_strategy_release(impl->retry_strategy);
+    aws_credentials_provider_release(impl->token_provider);
+
+    /* aws_http_connection_manager_release will eventually leads to call of s_on_connection_manager_shutdown,
+     * which will do memory release for provider and impl. So We should be freeing impl
+     * related memory first, then call aws_http_connection_manager_release.
+     */
+    if (impl->connection_manager) {
+        impl->function_table->aws_http_connection_manager_release(impl->connection_manager);
+    } else {
+        /* If provider setup failed halfway through, connection_manager might not exist.
+         * In this case invoke shutdown completion callback directly to finish cleanup */
+        s_on_connection_manager_shutdown(provider);
+    }
+}
+
+static struct aws_credentials_provider_vtable s_aws_credentials_provider_http_vtable = {
+    .get_credentials = s_aws_http_credentials_provider_get_credentials,
+    .destroy = s_credentials_provider_http_destroy,
+};
+
+int aws_http_credentials_provider_init_base(
+    struct aws_allocator *allocator,
+    struct aws_credentials_provider *provider,
+    struct aws_credentials_provider_http_options *options,
+    struct aws_http_credentials_provider_request_vtable *request_vtable) {
+
+    struct aws_credentials_provider_http_impl *impl =
+        aws_mem_acquire(allocator, sizeof(struct aws_credentials_provider_http_impl));
+
+    AWS_ZERO_STRUCT(*impl);
+    impl->request_vtable = request_vtable;
+
+    aws_credentials_provider_init_base(provider, allocator, &s_aws_credentials_provider_http_vtable, impl);
+
+    struct aws_tls_connection_options tls_connection_options;
```

Comment:
```suggestion
    struct aws_tls_connection_options tls_connection_options;
    AWS_ZERO_STRUCT(tls_connection_options);
```
Otherwise, when we do `aws_tls_connection_options_clean_up` from `on_error` couple lines below, it will be garbage.